fireant/queries/builder/dataset_query_builder.py

```diff
@@ -143,11 +152,109 @@ def fetch(self, hint=None) -> Iterable[Dict]:
+        annotation_alignment_dimension_filters = [
+            dataset_alignment_filter.for_(annotation.alignment_field)
+            for dataset_alignment_filter in dataset_alignment_dimension_filters
+        ]
+
+        annotation_alignment_field = annotation.alignment_field
+        if annotation_alignment_field.data_type == DataType.date:
+            dataset_alignment_dimension = self.fetch_query_dimension(
+                annotation.dataset_alignment_field_alias
+            )
+
+            if hasattr(dataset_alignment_dimension, "interval_key"):
+                # Use the interval key of the dataset's alignment dimension for the annotation's alignment field
+                # Otherwise we would need to copy it to prevent issues from patching directly
+                annotation_alignment_field = DatetimeInterval(
+                    annotation.alignment_field, dataset_alignment_dimension.interval_key
+                )
+
+        annotation_dimensions = [annotation_alignment_field, annotation.field]
+
+        annotation_query = make_slicer_query(
+            database=self.dataset.database,
+            base_table=annotation.table,
+            dimensions=annotation_dimensions,
+            filters=annotation_alignment_dimension_filters,
+        )
+
+        annotation_df = fetch_data(
+            self.dataset.database, [annotation_query], [annotation.alignment_field]
+        )
+
+        return annotation_df
+
+    def fetch_query_filters(self, dimension_alias):
+        """
+        Fetch all filters matching the given dimension alias from this query builder. All fields of a filter
+        (e.g. complex fields) have to match the alias and use the base table of the dataset.
+
+        :param dimension_alias:
+            An alias of a dimension.
+        :return:
+            A list of filters for the given dimension alias.
+        """
+        dimension_filters = []
+
+        for filter_ in self._filters:
+            filter_fields = [
+                field
+                for field in filter_.definition.fields_()
+                if all(table == self.dataset.table for table in field.tables_)
+            ]
+
+            if not filter_fields:
+                continue
+
+            if all(field.name == dimension_alias for field in filter_fields):
+                dimension_filters.append(filter_)
+
+        return dimension_filters
+
+    def fetch_query_dimension(self, dimension_alias):
+        """
+        Fetch a dimension matching the given dimension alias from this query builder.
+
+        :param dimension_alias:
+            An alias of a dimension.
+        :return:
+            A dimension (Field) of this query builder matching the dimension alias.
+        """
+        for dimension in self._dimensions:
+            unwrapped_dimension = dimension
+            while hasattr(unwrapped_dimension, "dimension"):
```

Comment:
I feel like I've seen this wrapping in multiple places. Is it possible to have all the wrapping and unwrapping functions in one place? I keep getting very confused by those.

Reply:
Good point, I've added a utility function for it.